### PR TITLE
Bf16 mixed precision

### PR DESCRIPTION
**Pull request by KangOxford**

### CODE REVIEW

**Review by KangOxford — COMMENTED** [review]
BF16 mixed precision, tensor core in the complex operations.

  What Actually Solved the Problem?

  Problem Evolution

  1. Initial problem: Training produced NaN
     - Loss = 17.47 million
     - Grad norm = 309.1 billion
     - Updated params has NaN

  2. First attempt: Full BF16 scan
     - Implemented BF16 associative_scan
     - Result: Numerical explosion xs in [-3e8, 3e8] -> NaN
     - Failure reason: BF16 insufficient precision, recursive error accumulation

  3. Second attempt: Data normalization
     - Ch0: /31 + clip to [-1, 1]
     - Vol: clip to [-1, 1]
     - Result: Loss dropped to 7.67, gradient 17.6
     - But still NaN: Updated params has NaN

  4. Third attempt: Rollback to FP32 scan
     - Scan uses FP32, Matmul uses BF16
     - Result: Forward/backward normal
     - But still NaN: tx.update() produces NaN

  5. Fourth attempt: Keep Lambda/D/log_step in FP32
     - Avoid BF16 precision issues with large param values (Lambda_im=-1303)
     - Result: Still NaN
     - Discovery: All params (including FP32 ones) become NaN

  6. Fifth attempt: Single optimizer (no multi_transform)
     - Result: Success! No NaN
     - Discovery: multi_transform has a bug

  7. Final solution: Initialize optimizer with FP32 + cast to FP32 on each update
     - Result: Success! No NaN, retained multi_transform functionality

  ---
  What Actually Solved It:

  Core Discovery

  When optax.multi_transform is initialized with BF16 params, its internal state
  records the BF16 dtype. This causes subsequent updates to produce NaN even when
  cast to FP32.

  The Real Solution

  Two key modifications:

  1. Initialize optimizer with FP32 (train_helpers.py:425-446)
     params_for_init = jax.tree_util.tree_map(to_fp32, params)  # temporary FP32
     state = TrainState.create(..., params=params_for_init, ...)  # initialize
     state = state.replace(params=params)  # swap back to BF16

  2. Cast to FP32 on each update (train_helpers.py:991-1045)
     params_fp32 = tree_map(to_fp32, state.params)
     grads_fp32 = tree_map(to_fp32, grads)
     updates, new_opt = state.tx.update(grads_fp32, opt_state, params_fp32)
     new_params_fp32 = apply_updates(params_fp32, updates)
     new_params = tree_map(cast_back, new_params_fp32, original_dtypes)

  ---
  Other Necessary But Insufficient Modifications

  These modifications were all necessary, but none alone could solve the problem:

  1. Data normalization - Reduced gradient from 309.1 billion -> 17.6
  2. FP32 scan - Fixed numerical explosion from +/-3e8 -> +/-3.9
  3. Lambda/D/log_step in FP32 - Avoided precision issues with large param values

  But without the final optimizer initialization + cast workaround, tx.update()
  would still produce NaN.

  ---
  Summary

  What actually solved the problem: Fixed multi_transform and BF16 interaction bug

  Method:
  1. Initialize optimizer with FP32 (avoid incorrect dtype recording)
  2. Cast to FP32 before each update (bypass dtype check)
  3. Cast back to original dtype after update (maintain BF16 storage)
